device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,205 @@
+    }
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_003: [When percentage is less than 0 or larger than 100, throw IllegalArgumentException.]
+    if (diagSamplingPercentage < 0 || diagSamplingPercentage > 100) {
+      throw new errors.ArgumentError('Sampling percentage should be [0,100]');
+    }
+    this.diagSamplingPercentage = diagSamplingPercentage;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_004: [This function shall reset message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.addDiagnosticInfoIfNecessary
+   * @description       add the diagnostic info
+   *
+   * @param {Message}   message   The message
+   */
+  addDiagnosticInfoIfNecessary(message: Message): void {
+    if (this.shouldAddDiagnosticInfo()) {
+      let diagnosticPropertyData: DiagnosticPropertyData = new DiagnosticPropertyData(this.generateEightRandomCharacters(), this.getCurrentTimeUtc());
+      message.diagnosticPropertyData = diagnosticPropertyData;
+    }
+  }
+
+  // Get a character from 0-9a-zA-Z
+  private getDiagnosticIdChar(value: number): string {
+    if (value <= 9) {
+      // 48 is char code of '0'
+      return String.fromCharCode(48 + value);
+    } else if (value <= 9 + 26) {
+      // 65 is char code of 'A'
+      return String.fromCharCode(65 + value - 10);
+    } else {
+      // 97 is char code of 'a'
+      return String.fromCharCode(97 + value - 36);
+    }
+  }
+
+  private generateEightRandomCharacters(): string {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_005: [This function shall generate 8 random chars, each is from 0-9a-z.]
+    let result: string = '';
+    for (let i = 0; i < 8; i++) {
+      let randomNum: number = Math.floor(Math.random() * DiagnosticClient.DIAGNOSTIC_ID_CHARACTER_BASE);
+      result += this.getDiagnosticIdChar(randomNum);
+    }
+    return result.toString();
+  }
+
+  private getCurrentTimeUtc(): string {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_006: [This function shall return the current timestamp in 0000000000.000 pattern.]
+    let currentDate = +new Date();
+    return (currentDate / 1000.0).toFixed(3);
+  }
+
+  private shouldAddDiagnosticInfo(): Boolean {
+    let result: Boolean = false;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_007: [This function shall return false if sampling percentage is set to 0.]
+    if (this.diagSamplingPercentage > 0 && this.diagSamplingPercentage <= 100) {
+      if (this.currentMessageNumber === 100) {
+        this.currentMessageNumber = 0;
+      }
+      this.currentMessageNumber++;
+      // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_008: [This function shall return value due to the sampling percentage setting.]
+      result = (Math.floor((this.currentMessageNumber - 2) * this.diagSamplingPercentage / 100.0) < Math.floor((this.currentMessageNumber - 1) * this.diagSamplingPercentage / 100.0));
+    }
+    return result;
```

Comment:
why not simply generate a random number and return true if <= diagSamplingPercentage?

Reply:
This is by our design. We would like to provide a stable rate of diagnostic sampling. If we use a random number, we can not ensure stability. Here's design doc about this: https://teams.microsoft.com/_#/docx/viewer/teams/https%3A~2F~2Fmicrosoft.sharepoint.com~2Fteams~2FIoTToolingTeam~2FShared%20Documents~2FEnd-to-End%20Diagnostics~2FAzure%20IoT%20Hub%20Device%20Client%20SDKs%20-%20E2E%20Diagnostic%20P0%20scenario.docx?threadId=19%3Acecb83aeeeea41bb818083fc32b545bd%40thread.skype&baseUrl=https%3A~2F~2Fmicrosoft.sharepoint.com~2Fteams~2FIoTToolingTeam&fileId=03F735E4-E751-4917-A3E4-EF04CFB2AAD6&ctx=files&viewerAction=view

Reply:
have you verified your assertion that using Math.random() doesn't give you a good enough distribution? how bad is the error? I'm going to reiterate: it's a pain to maintain this type of stuff and it's going to be prone to breaking. if we get good enough performance with simpler code, simpler should be the way to go.

---
In reply to: [155441804](https://github.com/Azure/azure-iot-sdk-node/pull/158#discussion_r155441804) [](ancestors = 155441804)